gunicorn/util.py

```diff
@@ -486,21 +486,28 @@ def daemonize(enable_stdio_inheritance=False):
             closerange(0, 3)
 
             fd_null = os.open(REDIRECT_TO, os.O_RDWR)
+            # PEP 446, make fd for /dev/null inheritable
+            os.set_inheritable(fd_null, True)
 
+            # expect fd_null to be always 0 here, but in-case not ...
             if fd_null != 0:
                 os.dup2(fd_null, 0)
 
             os.dup2(fd_null, 1)
             os.dup2(fd_null, 2)
 
         else:
-            fd_null = os.open(REDIRECT_TO, os.O_RDWR)
-
             # Always redirect stdin to /dev/null as we would
             # never expect to need to read interactive input.
 
+            os.close(0)
```

Comment:
why closing it there?

Reply:
This is making a best effort to create this new dev null as fd 0, and not creating an fd 3 that needs to be duplicated to 0. I think this is just trying to be more direct.

Reply:
It's also nice symmetry. In one case, we close 0-3, then make a new 0-3. In the other, we close only 0, and make a new 0.

If we wanted to, we could make the code a little shorter by closing and recreating fd 0 unconditionally. Then the if branches are just about what to do with fd 1 and 2.

Reply:
I think that might a actually be helpful for the reader to see that we never keep stdin open. Even though the option is to enable stdio inheritance, it really only allows stdout and stderr inheritance.

Reply:
Right, this is mostly for consistency with the fd-s created in both `not enable_stdio_inheritance` and `else`.  Since the aim is always to `</dev/null`, it's most direct to explicitly `close(0)` first in both cases, then `/dev/null` will (presumably always) be created as `fd=0`, which will skip the `dup2()` via the `if`, and the `dup2()` is a noop in that case anyway, which was the original cause of the non-inheritable fd raised in the issue.

Currently when `enable_stdio_inheritance`, `/dev/null` is created as `fd=3`, which is non-inheritable and not closed, so the worker ends-up with an extra `fd=3=/dev/null`, but any `subprocess` will only have `0/1/2`.

Since in both cases we always want to at least `close(0)` then create an inheritable replacement, it makes sense to me to do all those things explicitly, and the `dup2` isn't required, and is retained only for some weird situation if `close(0)` then `os.open()` created a non-zero fd, which shouldn't happen.

I can look at another commit to refactor a little more and make the stdin redirect common.

The other option of course is to not touch the `enable_stdio_inheritance` branch, since it currently fortuitously doen't have the non-inheritable stdin fd - that would mean ignoring the slight current difference, where with `enable_stdio_inheritance` the worker has an additional open, non-inheritable `fd=3=/edv/null`, which is mostly inconsequential.

Reply:
well it's a common practice to no close until you really need it. I would keep it that way and focus the commit on what really need to be fixed.

Reply:
here is also related old commit : https://github.com/benoitc/gunicorn/commit/29aefcc1cfc32097481aea6213a0df09962c7817

I would fix what latest Python has broken and have a separate patch/discussion about the rest.

Reply:
Thanks @benoitc for the review.

I've pushed a commit that reverts the refactoring change, and another that reverts the part of the original change to make the `-R` branch operate the same as the main branch i.e. `os.close(0)` then `os.open(...)` -> 0.  Running with `-R` will retain an additional `3 -> /dev/null`, non-inheritable, which is being relied-on (3 != 0) in order to trigger make-inheritable as part of `dup2(...).  As you say, these could be considered in a future PR if desired.

Retained is `os.set_inheritable(fd_null, True)` in the non `-R` branch, and some clarifying comments only.  This directly undoes the effect of PEP 446 in Python 3.4, which occurs due to the direct replacement of fd 0, bypassing `dup2(...)`.